Add a base tag mechanism and function template support to DataBox

nilsdeppe opened the pull request:
## Proposed changes

This adds 2 major and very closely related functionalities to DataBox:
- Compute items can now be function templates. This is done by having the
  function member be a function template, which then forwards to the free
  function.
- It is now possible to retrieve tags from the DataBox via a base tag
  mechanism. This is supported via two different styles: base tags and compute
  tags deriving off of a simple tag.
  For base tags the tag contains no members and must derive off of
  db::BaseDataBoxTags. It is then possible to retrieve the base tag MyBaseTag
  using db::get<MyBaseTag>(box). Note that it is also possible to mutate base
  tags (only if they are a simple item) by using db::mutate<MyBaseTag>(Box, ...)

  For compute items that derive off of simple tags the same approach applies,
  except that it is not possible to mutate the compute items (this rule hasn't
  changed)

Remaining TODOs:
- [ ] Write documentation
- [x] Rename ComputeItemTag to ComputeTag, DataBoxTag to SimpleTag
- [x] Clean up SimpleTag, ComputeTag, etc. inheritance

### Types of changes:

- [ ] Bugfix
- [x] New feature

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- Follows [code review guidelines](https://sxs-collaboration.github.io/spectre/code_review_guide.html)
- Code has documentation and unit tests
- Private member variables have a trailing underscore
- Do not use [Hungarian notation](https://en.wikipedia.org/wiki/Hungarian_notation), e.g. `double* pd_blah` is bad
- Header order:
  1. hpp corresponding to cpp (only in cpp files) or `tests/Unit/TestingFramework.hpp` (only in tests)
  2. Blank line (only in cpp files and tests)
  3. STL and externals (in alphabetical order)
  4. Blank line
  5. SpECTRE includes (in alphabetical order)
- File lists in CMake are alphabetical
- Correct `noexcept` specification for functions (if unsure, mark `noexcept`)
- Mark objects `const` whenever possible
- Almost always `auto`, except with expression templates, i.e. `DataVector`
- All commits for performance changes provide quantitative evidence and the tests used to obtain said evidence.
- Make sure error messages are helpful, e.g. "The number of grid points in the matrix 'F' is not the same as the number of grid points in the determinant."
- Prefix commits addressing PR requests with `fixup`. These commits are flagged
  by Travis so we do not accidentally merge them.
- Include what you use, prefer forward declarations in hpp files.
- Explicitly make numbers floating point, e.g. `2.` or `2.0` over `2`
- Use `Tensor`'s non-member get if the indices are constant expressions, e.g.
  `get<1>(tensor)`

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...
-->

nilsdeppe committed:
Add a base tag mechanism and function templates to DataBox

This commit adds 2 major and very closely related functionalities to DataBox:
- Compute items can now be function templates. This is done by having the
  function member be a function template, which then forwards to the free
  function.
- It is now possible to retrieve tags from the DataBox via a base tag
  mechanism. This is supported via two different styles: base tags and compute
  tags deriving off of a simple tag.
  For base tags the tag contains no members and must derive off of
  db::BaseDataBoxTags. It is then possible to retrieve the base tag MyBaseTag
  using db::get<MyBaseTag>(box). Note that it is also possible to mutate base
  tags (only if they are a simple item) by using db::mutate<MyBaseTag>(Box, ...)

  For compute items that derive off of simple tags the same approach applies,
  except that it is not possible to mutate the compute items (this rule hasn't
  changed)

kidder commented on tests/Unit/DataStructures/Test_BaseTags.cpp:
```diff
@@ -0,0 +1,336 @@
+struct ArrayBase : db::BaseDataBoxTag {};
+
+template <int I>
+struct Array : virtual db::DataBoxTag, ArrayBase<I> {
```
no virtual

nilsdeppe replied:
Turns out it's necessary otherwise GCC complains about it being an inaccessible base class

kidder commented on tests/Unit/DataStructures/Test_BaseTags.cpp:
```diff
@@ -0,0 +1,336 @@
+struct ArrayBase : db::BaseDataBoxTag {};
+
+template <int I>
+struct Array : virtual db::DataBoxTag, ArrayBase<I> {
```
DataBoxTag -> db::SimpleTag

kidder commented on tests/Unit/DataStructures/Test_BaseTags.cpp:
```diff
@@ -0,0 +1,336 @@
+namespace {
+namespace TestTags {
+template <int I>
+struct VectorBase : db::BaseDataBoxTag {};
```
db::BaseTag

kidder commented on tests/Unit/DataStructures/Test_BaseTags.cpp:
```diff
@@ -0,0 +1,336 @@
+};
+
+template <int I, int VectorBaseIndex = 0, int... VectorBaseExtraIndices>
+struct ArrayComputeBase : Array<I>, db::ComputeItemTag {
```
ComputeItemTag -> db::ComputeTag

kidder commented on tests/Unit/DataStructures/Test_BaseTags.cpp:
```diff
@@ -0,0 +1,336 @@
+  // - `get`ing an item by a pure base tag
+  // - `mutate`ing an item by a pure base tag
+  // - using a pure base tag as the argument in a compute item
+  // - `get`ing a compute item by pure base tag
```
remove pure and rename "impure" base to simple

kidder commented on tests/Unit/DataStructures/Test_BaseTags.cpp:
```diff
@@ -0,0 +1,336 @@
+  //   item and also a base tag
+  // - removing a compute item by its base tag
+  // - removing a compute item and its dependencies by the base tags
+  auto box = db::create<db::AddSimpleTags<TestTags::Vector<0>>,
```
intersperse comments with checks

kidder replied:
add global documentation to the DataBoxGroup and make these \snippets

kidder commented on tests/Unit/DataStructures/Test_BaseTags.cpp:
```diff
@@ -0,0 +1,336 @@
+  static constexpr db::DataBoxString label = "Parent";
+  static auto function(
+      const std::pair<Boxed<int>, Boxed<double>>& arg) noexcept {
+    count++;
```
remove count

kidder commented on tests/Unit/DataStructures/Test_BaseTags.cpp:
```diff
@@ -0,0 +1,336 @@
+        std::make_pair(TestTags::Boxed<int>(std::make_shared<int>(6)),
+                       TestTags::Boxed<double>(std::make_shared<double>(7.0))));
+
+  db::mutate<TestTags::FirstBase<0>>(
```
also test mutating the sub-item collection

kidder commented on tests/Unit/DataStructures/Test_DataBox.cpp:
```diff
@@ -1649,7 +1620,7 @@ SPECTRE_TEST_CASE("Unit.DataStructures.DataBox.Subitems",
 
   db::mutate<test_subitems::Second<0>>(
-      box, [](test_subitems::Boxed<double>& x) noexcept { *x = 12.; });
+      box, [](test_subitems::Boxed<double> & x) noexcept { *x = 12.; });
```
don't change

kidder commented on src/DataStructures/DataBox/DataBoxTag.hpp:
```diff
@@ -72,6 +72,25 @@ using DataBoxString = const char* const;
+ * off them for them to be useful
+ *
+ * Pure base tags do not need to even contain type information, unlike simple
+ * tags which must contain the type information. Pure base tags should be used
```
remove sparingly sentence

kidder commented on src/DataStructures/DataBox/DataBoxTag.hpp:
```diff
@@ -99,7 +118,7 @@ struct DataBoxTag {};
  * \see DataBox DataBoxTag DataBoxString get_tag_name ComputeItemTag
  */
-struct DataBoxPrefix : DataBoxTag {};
+struct DataBoxPrefix : virtual DataBoxTag {};
```
try to remove virtual

kidder commented on src/DataStructures/DataBox/DataBoxTag.hpp:
```diff
@@ -142,7 +161,19 @@ struct DataBoxPrefix : DataBoxTag {};
+    cpp17::is_same_v<Tag, ::Tags::DataBox>, tmpl::list<::Tags::DataBox>,
+    tmpl::filter<TagList, std::is_base_of<tmpl::pin<Tag>, tmpl::_1>>>;
+
+template <typename TagList, typename Tag>
```
put these 3 into detail namespace

kidder commented on src/DataStructures/DataBox/DataBoxTag.hpp:
```diff
@@ -305,7 +334,28 @@ struct is_compute_item<T,
+template <typename Tag>
+struct is_base_tag<Tag,
+                   Requires<cpp17::is_base_of_v<db::BaseDataBoxTag, Tag> and
+                            not cpp17::is_base_of_v<db::DataBoxTag, Tag>>>
```
and not compute tag

kidder commented on src/DataStructures/DataBox/DataBoxTag.hpp:
```diff
@@ -318,62 +368,107 @@ struct has_return_type_member<T, cpp17::void_t<typename T::return_type>>
+  // item_type_impl on the result.
+  template <typename TagList, typename Tag>
+  using f =
+      typename item_type_impl<TagList, first_matching_tag<TagList, Tag>>::type;
```
comment why no check for only one tag

kidder commented on src/DataStructures/DataBox/DataBoxTag.hpp:
```diff
@@ -318,62 +368,107 @@ struct has_return_type_member<T, cpp17::void_t<typename T::return_type>>
+      "compute item tag that is the problem should be shown in the first line "
+      " after the static assert error: "
+      "'detail::compute_item_result_helper<TheItemThatsFailingToBeCalled, "
+      "list<PassedTags...>>')");
```
fix this to show typelist of argument types and typelist of argument tags

kidder commented on src/DataStructures/DataBox/DataBoxTag.hpp:
```diff
@@ -318,62 +368,107 @@ struct has_return_type_member<T, cpp17::void_t<typename T::return_type>>
+struct compute_item_type<tmpl::list<Args...>> {
+  template <typename TagList, typename Tag>
+  using f = decltype(
+#ifdef SPECTRE_DEBUG
```
debug_static_assert

kidder commented on src/DataStructures/DataBox/DataBoxTag.hpp:
```diff
@@ -318,62 +368,107 @@ struct has_return_type_member<T, cpp17::void_t<typename T::return_type>>
+  template <typename TagList, typename Tag>
+  using f = decltype(
+#ifdef SPECTRE_DEBUG
+      check_compute_item_is_invokable<
```
(void)

kidder commented on src/DataStructures/DataBox/DataBox.hpp:
```diff
@@ -461,82 +516,299 @@ class TaggedDeferredTuple<> {
+// dep_items_from_list_impl is a left fold, but Brigand doesn't do folds through
+// aliases, so it's cheaper this way
+template <int FastTrackSelector>
+struct dep_items_impl;
```
dep_items -> expand_subitems

kidder commented on src/DataStructures/DataBox/DataBox.hpp:
```diff
@@ -461,82 +516,299 @@ class TaggedDeferredTuple<> {
-    tmpl::not_<std::is_same<typename Subitems<T>::type, tmpl::list<>>>;
+namespace detail {
+constexpr int select_dep_items_impl(const size_t pack_size) noexcept {
+  return pack_size >= 4 ? 3 : pack_size >= 2 ? 2 : static_cast<int>(pack_size);
```
comment what this is doing and implied of limit of items in DataBox

kidder commented on src/DataStructures/DataBox/DataBox.hpp:
```diff
@@ -461,82 +516,299 @@ class TaggedDeferredTuple<> {
+ * into a parameter pack so that we can use type aliases to do the computation,
+ * rather than structs. This turns out to be better for compilation speed.
+ *
+ * Once the `CompuetTagsList` is turned into a parameter pack, the
```
Compute

kidder commented on src/DataStructures/DataBox/DataBox.hpp:
```diff
@@ -461,82 +516,299 @@ class TaggedDeferredTuple<> {
+ *
+ * Once the `CompuetTagsList` is turned into a parameter pack, the
+ * `dep_items_impl` struct's nested type alias is used for the recursive
+ * computation. This recursive nature limits us to about 200 items in the
```
200 -> 800

kidder commented on src/DataStructures/DataBox/DataBox.hpp:
```diff
@@ -461,82 +516,299 @@ class TaggedDeferredTuple<> {
+};
+
+template <>
+struct expanded_list_from_full_list_impl<tmpl::list<>> {
```
can you be smarter about this?

kidder commented on src/DataStructures/DataBox/DataBox.hpp:
```diff
@@ -461,82 +516,299 @@ class TaggedDeferredTuple<> {
+using dep_items_from_list = typename detail::expanded_list_from_full_list_impl<
+    TagsList>::template f<FullTagList>;
+
+template <typename TagList, typename Tag>
```
why is this being added?

nilsdeppe replied:
It was moved into a detail namespace

nilsdeppe replied:
moved into a detail namespace

kidder commented on src/DataStructures/DataBox/DataBox.hpp:
```diff
@@ -581,22 +853,38 @@ class DataBox<TagsList<Tags...>> {
+  using simple_item_tags =
+      tmpl::list_difference<tags_list, compute_with_subitems_tags>;
+
+  /// A list of the simple items that are subitems, without expanding the
```
that are subitems -> that have subitems

kidder commented on src/DataStructures/DataBox/DataBox.hpp:
```diff
@@ -755,69 +1016,106 @@ class DataBox<TagsList<Tags...>> {
+
+template <>
+struct compute_item_function_pointer_type_impl<true> {
+  template <typename FullTagList, typename ComputeItem,
```
// mutating vs // non-mutating

kidder commented on src/DataStructures/DataBox/DataBox.hpp:
```diff
@@ -1081,34 +1404,41 @@ template <typename... MutateTags, typename TagList, typename Invokable,
+  static_assert(
+      not tmpl2::flat_any_v<
+          db::is_compute_item_v<first_matching_tag<TagList, MutateTags>>...>,
+      "Cannot mutate a compute item");
```
should check uniqueness

nilsdeppe committed:
Refactor DataBox completely

Simplify the DataBox implementation and make it better encapsulated.
- Remove TaggedDeferredTuple having DataBox itself have the leaves
- Rename get_lazy to get_deferred and make it a private method
- Have most of the methods in DataBox_detail that were used to create the
  DataBox be private methods of DataBox. This is required to make get_deferred a
  private method.
- As part of the complete change code is moved around so that the code is more
  logically located and easier to read. Comments delimiting logical groups of
  functions are added.

nilsdeppe committed:
Add DEBUG_STATIC_ASSERT

nilsdeppe committed:
ComputeItemTag and DataBoxPrefixTag don't inherit DataBoxTag

This simplifies some of the logic in the implementation and means that
PrefixTag must be matched with either ComputeTag or SimpleTag (currently still
named DataBoxTag)

nilsdeppe committed:
Rename db::DataBoxPrefix to db::PrefixTag

nilsdeppe committed:
Rename db::DataBoxTag to db::SimpleTag

nilsdeppe committed:
Rename db::ComputeItemTag to db::ComputeTag

nilsdeppe committed:
Add rvalue reference overload for db::create_from

nilsdeppe committed:
Reset Subitems compute items more efficiently.

Previously a new Deferred was created (resulting in allocations) each time a
subitem was changed. Now we only reevaluate the compute items.

nilsdeppe committed:
Optimize Subitems<Variables>

Previously the pointers inside the Tensors are added to the DataBox from the
Variables and would be updated each time the Variables was changed. Now we
check if the allocation changed and only then update the pointers.

nilsdeppe committed:
Document the DataBox

nilsdeppe committed:
Make db::mutate take a gsl::not_null<DataBox*>

nilsdeppe committed:
Make db::mutate_apply take db::DataBox by gsl::not_null

nilsdeppe committed:
Remove db::apply_with_box since it is not needed

We can retrieve the DataBox from the DataBox using the Tags::DataBox tag.

nilsdeppe committed:
Rename db::DataBoxString to db::Label

kidder commented on src/DataStructures/DataBox/DataBox.hpp:
```diff
@@ -859,25 +810,25 @@ class DataBox<TagsList<Tags...>> {
   /// Specifically, if there is a `Variables<Tag0, Tag1>`, then this list would
   /// contain `Tag0, Tag1`.
-  using simple_only_expanded_subitems_tags = tmpl::flatten<tmpl::transform<
+  using simple_only_expand_subitems_tags = tmpl::flatten<tmpl::transform<
```
undo this

kidder commented on src/DataStructures/DataBox/DataBoxTag.hpp:
```diff
@@ -74,22 +74,21 @@ struct DataBoxTag {};
- * sparingly, and are designed only so that retrieving items from the DataBox or
- * setting argument tags in compute items can be done without any knowledge of
- * the type of the item.
+ * Base tags do not need to even contain type information, unlike simple
```
remove even

kidder commented on src/DataStructures/DataBox/DataBoxTag.hpp:
```diff
@@ -74,22 +74,21 @@ struct DataBoxTag {};
  *
- * To use the pure base mechanism the pure base tag must inherit off of
- * `BaseDataBoxTag` and NOT `DataBoxTag`. This is very important for the
+ * To use the base mechanism the base tag must inherit off of
```
remove "To ... mechanism"

kidder commented on src/DataStructures/DataBox/DataBoxTag.hpp:
```diff
@@ -431,32 +433,44 @@ template <>
-      "list<PassedTags...>>')");
+      "'check_compute_item_is_invokable<TheItemThatsFailingToBeCalled, "
+      "brigand::list<PassedTags...>, "
+      "brigand::list<item_type<Passedtags>...>>')");
```
remove first )

kidder replied:
PassedTags

kidder commented on src/DataStructures/DataBox/DataBox.hpp:
```diff
@@ -93,484 +83,51 @@ class TaggedDeferredTupleLeaf {
-               cpp17::is_constructible_v<value_type, T&&>> = nullptr>
-  constexpr explicit TaggedDeferredTupleLeaf(T&& t) noexcept(  // NOLINT
+  template <class T,
+            Requires<!cpp17::is_same_v<std::decay_t<T>, DataBoxLeaf> &&
```
ands and nots

kidder commented on src/DataStructures/DataBox/DataBox.hpp:
```diff
@@ -1145,193 +697,279 @@ add_compute_item_to_box_impl(databox_detail::TaggedDeferredTuple<Tags...>& data,
+db::DataBox<tmpl::list<Tags...>>::add_subitem_tags_to_box(
+    tmpl::list<Subtags...> /*meta*/) noexcept {
+  const auto helper = [this](auto tag_v) {
+    (void)this;  // Compiler bug wars this unused
```
warns

kidder commented on src/DataStructures/DataBox/DataBox.hpp:
```diff
@@ -1145,193 +697,279 @@ add_compute_item_to_box_impl(databox_detail::TaggedDeferredTuple<Tags...>& data,
+  };
+
+  static_cast<void>(std::initializer_list<char>{
+      (static_cast<void>(helper(Subtags{})), '0')...});
```
write the macro

kidder commented on src/DataStructures/DataBox/DataBox.hpp:
```diff
@@ -1145,193 +697,279 @@ add_compute_item_to_box_impl(databox_detail::TaggedDeferredTuple<Tags...>& data,
+constexpr DataBox<tmpl::list<Tags...>>::DataBox(
+    const DataBox<OldTags>& old_box, tmpl::list<KeepTags...> /*meta*/,
+    tmpl::list<AddTags...> /*meta*/, tmpl::list<AddComputeTags...> /*meta*/,
+    Args&&... args) {
```
noexcept

kidder commented on src/DataStructures/DataBox/DataBox.hpp:
```diff
@@ -1145,193 +697,279 @@ add_compute_item_to_box_impl(databox_detail::TaggedDeferredTuple<Tags...>& data,
-  add_sub_compute_item_tags_to_box<ComputeItem>(
-      data, typename Subitems<tmpl::list<Tags...>, ComputeItem>::type{});
-}
+  // Merge old tags, including all ComputeItems even though they might be
```
remove the comment

kidder commented on src/DataStructures/DataBox/DataBox.hpp:
```diff
@@ -1614,13 +1614,16 @@ constexpr bool check_mutate_apply_argument_tags(
-                                          std::forward<Args>(args)...))) {
+inline constexpr auto mutate_apply(
+    F f, const gsl::not_null<DataBox<BoxTags>*> box,
+    Args&&... args) noexcept(DataBox_detail::
```
kill this noexcept and the comment about it

nilsdeppe commented on docs/GroupDefs.hpp:
```diff
@@ -119,7 +119,214 @@
+ * \brief Documentation, functions, metafunctions, and classes necessary for
+ * using DataBox
+ *
+ * DataBox is a heterogeneous compile-time hash map with lazy evaluation of
```
hash map -> associative container

nilsdeppe commented on docs/GroupDefs.hpp:
```diff
@@ -119,7 +119,214 @@
+ *
+ * #### Tags and Items
+ *
+ * The compile-time keys are `struct`s called Tags, while the values are called
```
Tags -> tags (throughout)

nilsdeppe commented on docs/GroupDefs.hpp:
```diff
@@ -119,7 +119,214 @@
+ *
+ * Simple tags must have a member type alias `type` that is the type of the data
+ * to be stored, a `static constexpr db::Label label` that is the name of the
+ * tag, and inherit from `db::SimpleTag`. Compute tags must also have a `static
```
New sentence at `, and`
New paragraph before Compute tags

nilsdeppe commented on docs/GroupDefs.hpp:
```diff
@@ -119,7 +119,214 @@
+ *
+ * \snippet Test_DataBox.cpp compute_item_tag_function
+ *
+ * Finally, a compute tag with an overloaded function and a function template
```
Add examples of:
- Overload on the type
- Overload on the number of args
- Simple function template like ComputeLambda0
- Crazy example

nilsdeppe commented on docs/GroupDefs.hpp:
```diff
@@ -119,7 +119,214 @@
+ * A simple or compute tag might also hold a collection of data, such as a bunch
+ * of `Tensor`s. In many cases you will want to be able to retrieve individual
+ * elements of the collection from the DataBox without having to first retrieve
+ * the collection. The infrastructure that allows for this is call *Subitems*.
```
`call` -> `called`

nilsdeppe commented on docs/GroupDefs.hpp:
```diff
@@ -119,7 +119,214 @@
+ *
+ * \snippet Test_BaseTags.cpp compute_template_base_tags
+ *
+ * A simple or compute tag might also hold a collection of data, such as a bunch
```
bunch -> container

nilsdeppe commented on docs/GroupDefs.hpp:
```diff
@@ -119,7 +119,214 @@
+ * added to the DataBox, so are `Sub<0>` and `Sub<1>`. This means the retrieval
+ * mechanisms described below will work on `Parent`, `Sub<0>`, and `Sub<1>`.
+ *
+ * Subitems are controlled not by the tag, but by the type of the tag. Let's say
```
Explain that this is done by a requirement on the input tag

nilsdeppe commented on docs/GroupDefs.hpp:
```diff
@@ -119,7 +119,214 @@
+ * See the documentation of `db::Subitems` for more details.
+ *
+ * The DataBox also supports *prefix tags*, which are commonly used for items
+ * that are related to a different item by a linear operator. Specifically, say
```
by a linear operator -> by some operation

nilsdeppe commented on docs/GroupDefs.hpp:
```diff
@@ -119,7 +119,214 @@
+ * to use `const auto&` when retrieving tags from the DataBox. For example,
+ * \snippet Test_DataBox.cpp using_db_get
+ *
+ * If you want to mutate the value of an item in the DataBox use `db::mutate`.
```
an item -> a simple item

nilsdeppe commented on docs/GroupDefs.hpp:
```diff
@@ -119,7 +119,214 @@
+ * Any compute item that depends on the mutated item will have its cached value
+ * invalidated and be recomputed the next time it is retrieved from the DataBox.
+ * `db::mutate` takes a parameter pack of tags to mutate as explicit template
+ * parameters, a `gsl::not_null` of the DataBox whose items to mutate, an
```
to mutate -> will be mutated

nilsdeppe commented on docs/GroupDefs.hpp:
```diff
@@ -119,7 +119,214 @@
+ * `db::get<Array>(box)`. The base tags mechanism requires that only one `Array`
+ * tag be present in the DataBox, otherwise a static assertion is triggered.
+ *
+ * #### The Base Tags Mechanism
```
move header above

nilsdeppe commented on docs/GroupDefs.hpp:
```diff
@@ -119,7 +119,214 @@
+ * #### The Base Tags Mechanism
+ *
+ * The inheritance idea can be generalized further with what are called a base
+ * tag. A base tag is an empty `struct` that inherits from `db::BaseTag`. Any
```
tag. -> tags.

nilsdeppe commented on docs/GroupDefs.hpp:
```diff
@@ -119,7 +119,214 @@
+ * including mutating `Vector<0>`, and then verifying that the dependencies are
+ * handled correctly.
+ *
+ * \snippet Test_BaseTags.cpp base_simple_and_compute_mutate
```
add white space in example

nilsdeppe replied:
Add comments to the example to highlight the different CHECK calls

nilsdeppe commented on docs/GroupDefs.hpp:
```diff
@@ -119,7 +119,214 @@
+ *
+ * \snippet Test_BaseTags.cpp base_simple_and_compute_mutate
+ *
+ * Notice that we are ably to retrieve `ArrayComputeBase<0>` with `ArrayBase<0>`
```
ably -> able

nilsdeppe commented on docs/GroupDefs.hpp:
```diff
@@ -119,7 +119,214 @@
+ * arguments are set in the member type alias `argument_tags`, which must be a
+ * `tmpl::list` of the tags corresponding to each argument. Note that the order
+ * of the tags in the `argument_list` is the order that they will be passed to
+ * the function. Compute tags must inherit from `db::ComputeTag`.
```
compute items must take `const &`

kidder reviewed (changes requested):
Cornell group code review

nilsdeppe committed:
Add EXPAND_PACK_LEFT_TO_RIGHT macro

Allows easy expansion of parameters packs guaranteeing left-to-right
evaluation.

nilsdeppe committed:
Stop tapping homebrew/science

homebrew/science is semi-deprecated with most of the packages there moved into
core homebrew, including HDF5, so we no longer need it.
